### ISSUE TITLE
Refactor Dependency Injection

### ISSUE THREAD

**Issue by brianalbrecht** [CenturyLinkCloud/clc-java-sdk#90]
The SDK (especially the ClcSdk class) is pretty tightly integrated with Google Guice.  We use Spring internally, and we'd like to see the project refactored to move away from Guice dependence and allow for use with Spring.  Overall, Spring is more mature and more widely adopted in enterprises, which will be a major consumer of this SDK.

Ideally, we would split this into two projects:  The current SDK project would become a service provider with no wiring/injection.  Then we would build a Spring injection project to wire the core services into a Spring context, and a Guice injection project to wire the core services into a Guice context.

Let me know what you think.

**Comment by jruckle** [comment]
@idrabenia - can you document how this configuration would work, now that it is close to starting?

**Comment by idrabenia** [comment]
Hi @jruckle! 

Injection to Spring DI managed code will be seems like. We will provide possibilities to inject all our Guice services to Spring managed code.

``` java
    @Autowired
    ServerService serverService;

    @Autowired
    GroupService groupService;

    @Autowired
    TemplateService templateService;

    @Autowired
    DataCenterService dataCenterService;

    @Autowired
    CredentialsProvider credentialsProvider;

    @Autowired
    StatisticsService statisticsService;
```

Bean that customer will need to provide in Spring configuration will be looks like
``` java
@Configuration
class MyConfig {
   
   @Bean
   public ClcSdkConfiguration springAdapter () {
      return new ClcConfiguration(
        /*CredentialsProvider*/ credentialsProvider, /*SdkConfiguration*/ config
      );
   }

}
```

End-user will need to provide only SDK configuration - results beans will be configuration implicitly by provided in SDK Spring configuration.

It will be possible to add this Spring adapter as separate JAR file.


If you will have some other questions - feel free to contact me or @pslam .

Have a good day!

**Comment by jruckle** [comment]
@brianalbrecht  - what do you think about the above from @idrabenia ?

**Comment by brianalbrecht** [comment]
@idrabenia @jruckle This will work, but it adds an extra, unnecessary dependency on Guice for anyone who wants to use the project.  That's not going to be popular.  Had this work started, and is it available on a feature branch?

**Comment by markramach** [comment]
I would consider eliminating Guice references like `@Inject` from the service classes all together, moving all configuration and injection to the Guice Module classes. That would allow you to move all the Module configuration classes to an alternate project. You could then have a similar configuration project for Spring.

**Comment by brianalbrecht** [comment]
I agree with @markramach.  Refactoring this way would make the sdk much more usable and sustainable.

**Comment by pslam** [comment]
Hello guys.
@brianalbrecht I've pushed spring adapter into feature-150 branch
@jruckle @brianalbrecht Should we implement another refactoring proposed @markramach ?

**Comment by jruckle** [comment]
@markramach @brianalbrecht  - where do you advise the team to go from there?  SHould this even be a seperate SDK or a fork?

**Comment by idrabenia** [comment]
Hi @brianalbrecht @jruckle @markramach @pslam!

About Google Guice - I think it's very lightweight solution and it do not require too much dependencies to download.

Also before start of SDK development we discussed DI with @nstetich , @jruckle and @scottdensmore . And we decided that Guice will be suitable solution for our SDK. I said that it would be difficult to change this core design aspect in future. All people was agreed about Guice usage in DI.

I think now it's too late to make such significant changes because we have strong dependency to Guice in all services, also our test framework based on Guice DI.

I think changing of this base aspect of architecture may require 3-4 man-month. And also we have high risks of failure for this task.

From my opinion Spring Adapter (which proposed in this ticket) - is best compromise in current situation.

But if you really want to make any changes of architecture and you think that it's really significant for you. We are open to discussion.

Have a good day :)
Ilya

**Comment by markramach** [comment]
@jruckle @brianalbrecht I think it really depends on how you want to differentiate actual SDK code from framework configuration code. If the current SDK project were to proceed as a nice clean stable set of model, client and service classes, then it would probably make sense to create separate projects for Spring and Guice framework configuration. From the look of it that would require a significant refactor per @idrabenia.

**Comment by idrabenia** [comment]
@brianalbrecht @jruckle @markramach @pslam

I think solution - to remove all `@inject` annotation and create two separate project possible will be very complex. We will need to test **two different environments** and make **dangerous core modification**. 
Also I did not see such solutions in other SDKs and at all. Usual solution - is adapter for other DI framework or manual DI management.

About Google Guice - it used as base DI in **Apache JClouds**. That is one of the best Java-libraries for manage cloud infrastructure. For integration with Spring it provide Spring adapters - similar solution that we proposed.

**Comment by idrabenia** [comment]
@brianalbrecht @jruckle @markramach @pslam

Than from my opinion - if you really want to reduce dependency to Google Guice - it would be best choice to manage dependencies manually. Because we will have one SDK, one environment and we will not rely on underlined realisation of DI. But I think this solution will have huge time effort at current stage of project.

**Comment by jruckle** [comment]
@idrabenia can you advise on how long it would take to re-factor the SDK to support spring as @markramach & @brianalbrecht have suggested earlier on this issue?

**Comment by pslam** [comment]
Hi guys
@jruckle as I see from @idrabenia comment above
It may require 3-4 man-month. And also we have high risks of failure for this task

**Comment by pslam** [comment]
@brianalbrecht @markramach
Could you describe in details which changes we should implement within your proposed refactoring ?

**Comment by idrabenia** [comment]
Hi @brianalbrecht @markramach @jruckle!

1. I think idea to create two separate projects will require *2 man-month* to implementation with high risk. 
**But if not difficult could you explain - which benefits you will expect if we will start our such way implementation. ** You want only reduce dependency to Guice - or you have ideas to provide pluggable modular SDK structure for each DI framework. 

2.

**Comment by markramach** [comment]
I'm really just talking about swapping out declaration for configuration. Removing the `@Inject` annotations and using constructors or getter/setters for configuration via the Module classes.

For example using this in the AuthModule

    @Provides
    public LoginClient getLoginClient(SdkConfiguration config) {
        return new LoginClient(config);
    }

Instead of this in the LoginClient type

    @Inject
    public LoginClient(SdkConfiguration config) {
        super(config);
    }

This turns the AuthClient into a simple pojo.

**Comment by cljvf9** [comment]
Hi all, is there any idea when a new maven release will be published with this new feature included? I would like to start using the sdk to perform some tasks that I currently do by writing specific API calls, but have been waiting to be able to integrate it with the rest of my project using spring.

**Comment by idrabenia** [comment]
Hi @cljvf9 @jruckle,
I created task to this update #160 
During 2-3 days we will publish new artefacts to maven central.

Best Regards,
Ilya

**Comment by idrabenia** [comment]
Hi @cljvf9 @jruckle,

We are completed SDK refactoring and published artefacts to maven central - 
http://search.maven.org/#search%7Cga%7C1%7Ccenturylink 

If you will have any questions related to SDK usage - will be glad to help you!

Best Regards,
Ilya

**Comment by cljvf9** [comment]
Cool, glad to hear it! I specified the new version 1.1.0 as a dependency in my project and started writing some code to see how I would use it to accomplish what I need -- I am running into trouble using the spring adapter and injecting my own CredentialsProvider (I want to use a StaticCredentialsProvider). Here is my configuration class:

```
@Configuration
@EnableClcSdk
public class ClcConfig {

    @Bean
    public CredentialsProvider clcCredentialsProvider() {
        return new StaticCredentialsProvider("username", "password");
    }

    @Bean
    public SdkConfiguration clcSdkConfig() {
        return new SdkConfigurationBuilder().build();
    }

}
```

Here is a throwaway test I wanted to use to see how I would accomplish something using the SDK (in this example creating a group hierarchy):

```
@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration(classes = ClcConfig.class)
public class CreateGroupHierarchyTest {

    @Autowired
    private GroupService groupService;

    @Test
    public void createGroupHierarchy() throws Exception {
        groupService.defineGroupHierarchy(DataCenter.refById("va1"), new GroupHierarchyConfig().name("first")
                .subitems(
                        new GroupHierarchyConfig().name("second")
                                .subitems(
                                        new GroupHierarchyConfig().name("third")
                                )
                )
        ).waitUntilComplete();
    }
}
```

However when I try to run it I keep getting `ClcException: Failed to provide credentials`. I think it is because the ClcSpringConfig class is constructing an instance of the DefaultCredentialsProvider before any injection takes place, so it is looking for credentials in the default locations even though I don't want it to. Can you confirm my suspicion, or am I doing something wrong? Should this be opened as a separate issue?

I am not sure how useful this piece of information is, but I am using spring version 4.2.0.RELEASE.

Thanks for the support

**Comment by jruckle** [comment]
@idrabenia - can you take a look at this issue? I'd like to get the Spring DI working before spending time on additional capabilities.

**Comment by idrabenia** [comment]
Hi @jruckle @cljvf9 ,

Thanks for your feedback!

We are fixed this bug. New SDK version now available in maven central.
http://repo1.maven.org/maven2/com/centurylink/cloud
 
*Important! We renamed **clc-sdk-spring-adapter** to **spring-clc-sdk** *

**Comment by cljvf9** [comment]
Thanks @idrabenia , I am now able to use a different CredentialsProvider in my spring config. One more thing I wanted to point out is when I declare the new spring-clc-sdk dependency it produces an error when trying to download its transitive dependencies, specifically the sdk:

```
<dependency>
  <groupId>clc-java-sdk</groupId>
  <artifactId>clc-java-sdk</artifactId>
  <version>1.1.1</version>
  <scope>compile</scope>
</dependency>
```
The groupId appears to be wrong (should it be `com.centurylink.cloud`?). I was able to work around this by excluding that and explicitly including the correct groupId:

```
<dependency>
    <groupId>com.centurylink.cloud</groupId>
    <artifactId>clc-java-sdk</artifactId>
    <version>1.1.1</version>
</dependency>
<dependency>
    <groupId>com.centurylink.cloud</groupId>
    <artifactId>spring-clc-sdk</artifactId>
    <version>1.1.1</version>
    <exclusions>
        <exclusion>
            <groupId>clc-java-sdk</groupId>
            <artifactId>clc-java-sdk</artifactId>
        </exclusion>
    </exclusions>
</dependency>
```

**Comment by idrabenia** [comment]
Hi @cljvf9 ,

Thanks for your feedback! 

Exactly, this is issue of SDK publishing script. It fixed and this artefacts with correct pom available as version **1.1.2**.

**Comment by cljvf9** [comment]
Thanks for the quick response, I updated my dependencies and it looks good.
